### DISCUSSION TITLE
NFR - Protect and encrypt sensitive data when stored persistently with strict data separation

### DISCUSSION THREAD

**Discussion by DanielaWuensch** [eclipse-edc/Connector#1675] [Ideas]
# Feature Request
Non-functional Requirement: 
Protect and encrypt sensitive data when stored persistently (with strict data separation between the sensitive data and non-sensitive data and data from different customers)
As a company, which operates the EDC I want to ensure that sensitive data is stored securely (encrypted). Currently the persistence is used in the data management api and the data is not encrypted. 
Highest priority would have the asset->dataAddress as it contains credentials of the backend service. These credentials should be protected with VH priority. Credentials should not be stored insecure and perfect place for it would be the vault.
Extract the the dataAddress in data mgt api and store it in the vault (which ensures secure persistence)
In addition all other data persisted should be encrypted with a key provided by the admin.
encrypt the db data

## Which Areas Would Be Affected?
all, including DPF, CI, build, transfer, etc._

## Why Is the Feature Desired?
Security Requirement

## Solution Proposal
sensitive data need alway to be stored securely (encrypted). 

## Type of Issue
non-functional requirement

## Checklist
all sensitive data are stored securely (encrypted).

**Comment by jimmarino** [comment]
The data address does not contain any secrets (this appears to be a misinterpretation of the code). All secrets are, by EDC design, stored in the Vault.

**Reply by ndr-brt** [reply]
I think we currently miss a way to have asset/dataaddress sensitive properties stored in the vault.
One solution could be to have a `secret:` prefix for the property key, for example:
```
{
  "properties: {
    ...
    "secret:api-token": "<vault_key>"
    ...
  }
}
```
This would mean that the `api-token` property it's stored in the vault with the `<vault_key>` key

**Reply by jimmarino** [reply]
I see the HttpDataAddress does this, which it shouldn't. Other uses correctly handle this. The keys should be sent outside the DataAddress and only the key name stored there.

**Reply by ndr-brt** [reply]
> I see the HttpDataAddress does this, which it shouldn't. Other uses correctly handle this. The keys should be sent outside the DataAddress and only the key name stored there.

Agree, my suggestion was to have a way to make the sink/source factories aware of the fact that they need to extract the value from the vault, using a `secret:` would be pretty neat

**Reply by lholthof** [reply]
In case you have multiple elements in the DataAddress which you do not want to expose to the consumer, it would be beneficial to have one vault-key and perform only one lookup instead of multiple if you would prefix several properties with `secret:`.
I am thinking of putting multiple properties, e.g. the authCode and the URL, into the vault as you could consider the internal URL also as a sensitive information revealing internal structures. But also additional headers or paths might be better put into the vault.

Therefore I suggest to use _secret:_, but not provide the vault-key value in it, instead use a vault-key by convention. The asset-id is unique and there will be one DataAddress per Asset. Hence this convention could be vault-key => <configurable_prefix>.<asset-id> (e.g. edc.assetsecret.XYZ).
Now in the Data Mgt. Asset API you can still perform one call to create the asset including _secret:_ prefixed properties. Those will be stored in the vault (e.g. serialized json). Reading an asset gets those values from the vault again and shows just the real values as a response. Side-remark: Ideally user authorization would influence if one can see those _secret:_ prefixed properties or not. Maybe that could be a future increment to either show the real values or mask them.

Would this be feasible?

On the dataplane repeated access to the vault content will be a usual pattern. On which level should the "costly" lookup to the vault be cached?

**Reply by jimmarino** [reply]
An important note: `DataAddress` functions similarly to a URL in the EDC, so we should be careful to highlight in which context they are used.

My preference would be to reflect the fact that they are secrets in the types that are used by the system instead of prefixes. On ingress, code will know implicitly which headers are secrets and can handle those accordingly by setting a different set of properties (and possibly types). Downstream code can then use those properties to resolve values from the Vault. The provisioning infrastructure does this, and we should align the use of HttpDataAddress with that.

IMO having a key in the data address is not leaking secrets. The vault address (URL) is never given out so the key is no different than a correlation id that is used to lookup entries such as a `TransferProcess` in a database. This information is also encrypted when it is sent to the client. There is no way to access the secret unless someone breaks the encryption, figures out the Vault URL used by the EDC deployment, obtains the Vault access credentials, and has network access to the Vault. This person would have compromised the entire EDC and network environment, which is not possible with just the DataAddress.

We could investigate performance based on concrete requirements. A lot of this can likely be solved by enabling caching on a Vault implementation client. Combining multiple keys into one payload could be done by individual parts of the code that are performance-sensitive, but that should only be done for temporary keys. IMO configuration secrets should be managed separately in the vault and not as part of a single JSON structure. Another thing to keep in mind is that most of the code is not latency sensitive due to the asynchronous system architecture. The data plane will be when proxying requests, and we have discussed introducing a high-performance validation and caching mechanism when needed.    

It appears there are two issues that need to be looked into:

1. Creating a way (preferably typed) to ensure secret values are not passed in `HttpDataAddress` (other `DataAddress` usages correctly do this). 

2. Address any performance-related issues in the context of a broader discussion about data plane contract validation and concrete requirements.

**Reply by lholthof** [reply]
If I get you right, then your suggestion would lead to a two step asset creation. 
Create vault secrets plus use the corresponding vault-keys in POST /asset.

I also do not see an issue in having a vault key in the data address. I just wanted to use it by convention to keep the convenience as-is for the data mgt API users, to have one call creating an asset including all details of the dataAddress and not to care about vault-keys.
With the prefixing the API user would have the choice to define which properties should be put into the vault. At the same time this leave a point of failure. Therefore why not combine both? Typed properties will be put into the vault and in addition the ones prefixed with _secret:_. Some will consider the URL as sensitive data and some might not agree to that. Like this everyone has the choice while we will ensure that e.g. an authCode (typed as Secret) will always be put into the vault.

Maybe one improvement on using the vault-key by convention (based on asset-id) would be to hash the content and use this hash as vault-key. The vault-key would be an additional property in the dataAddress. With this repeated same secrets will not consume an individual vault entry all the time. At the same time the items in the vault cache will be less which improves performance.

**Reply by jimmarino** [reply]
I think we are introducing too many topics into the discussion at once and this is leading to confusion. So far, we have one actionable issue:

1.  Creating a way (preferably typed) to ensure secret values are not passed in HttpDataAddress (other DataAddress usages correctly do this)

Performance was also mentioned but I would suggest putting that aside for the purposes of this discussion since it is a broader issue. 

Another point that is raised relates to the creation of an `Asset` and its associated `DataAddress` (this is distinct from the actionable issue since data addresses are used throughout the codebase in diverse contexts for dereferencing assets). The Vault secret should be generated out-of-band and stored in the Vault when the asset contents are stored at the location referenced by the data address (which is *not* necessarily when the `Asset` and `DataAddress` are created in the EDC). Users entering assets should not have access to those secrets nor should they populate them in the Vault. This is a separate topic, though.

**Reply by dileep-ramachandrarao** [reply]
Following the thread above, I have few clarifications and for the purpose of better understanding, I would like to discuss using a simple example.

Currently, during Asset creation if the data endpoint is password protected then the data address would look like this:
```
"dataAddress": {
        "properties": {
            "type": "HttpData",
            "endpoint": "http://provider-backend-service:9191/data/{{digitalTwinSubmodelId}}",
	    "authCode": "Basic an7uthklGywTMnMkM="
        }
    }

```
Here `authCode` is a sensitive data, which contains Base64 encoded username and password and it is persisted without encryption.

As suggested in the above thread, prefix the sensitive properties using `secret:` and store them in the vault.
By following further discussion, I see that we could either 

_Option 1_: hash all the sensitive properties and store them in the vault like:
```
vaultContent={"secret:endpoint": "http://provider-backend-service:9191/data/{{digitalTwinSubmodelId}}",
"secret:authCode": "Basic an7uthklGywTMnMkM="}
hash(vaultContent)= h849dkckkd //would be the vault key
```

Within the vault, the sensitive properties could be accessed using the vault key:
```
VaultKey: h849dkckkd
VaultContent: {"secret:endpoint": "http://provider-backend-service:9191/data/{{digitalTwinSubmodelId}}",
"secret:authCode": "Basic an7uthklGywTMnMkM="}

```
Following would be stored in the DB:
```
"dataAddress": {
    "properties": {
      "secret:endpoint": "h849dkckkd",
      "type": "HttpData",
      "secret:authCode": "h849dkckkd"
    }
  }
```

OR

_Option 2_: hash of the respective sensitive property:
```
vaultContent={"secret:endpoint": "http://provider-backend-service:9191/data/{{digitalTwinSubmodelId}}"}
hash(vaultContent)= 123564987
VaultKey: 123564987

vaultContent={"secret:authCode": "Basic an7uthklGywTMnMkM=}
hash(vaultContent)= abcdklskdjf
VaultKey: abcdklskdjf
```

Following would be stored in the DB:
```
"dataAddress": {
    "properties": {
      "secret:endpoint": "123564987",
      "type": "HttpData",
      "secret:authCode": "abcdklskdjf"
    }
  }
```

Does either of the approaches a right way to address this issue?

**Reply by dileep-ramachandrarao** [reply]
Trying to revive this conversation. @jimmarino @ndr-brt any opinion on this?

**Reply by jimmarino** [reply]
Putting sensitive data in the `DataAddress` is incorrect. If code is doing this, it needs to be fixed. The only data that should be placed in the data address is non-sensitive data and pointers to where sensitive data may be retrieved from a vault.

There is no need to hash or encrypt sensitive data in a vault since the vault implementation handles that.

HTH.
